tsfc/loopy.py

```diff
@@ -0,0 +1,338 @@
+            else:
+                return pym
+        else:
+            return pym
+
+    def active_inames(self):
+        # Return all active indices
+        return frozenset([i.name for i in self.active_indices.values()])
+
+
+def generate(impero_c, args, precision, kernel_name="loopy_kernel", index_names=[]):
+    """Generates loopy code.
+
+    :arg impero_c: ImperoC tuple with Impero AST and other data
+    :arg args: list of loopy.GlobalArgs
+    :arg precision: floating-point precision for printing
+    :arg kernel_name: function name of the kernel
+    :arg index_names: pre-assigned index names
+    :returns: loopy kernel
+    """
+    ctx = LoopyContext()
+    ctx.indices = impero_c.indices
+    ctx.index_names = defaultdict(lambda: "i", index_names)
+    ctx.precision = precision
+    ctx.epsilon = 10.0 ** (-precision)
+
+    # Create arguments
+    data = list(args)
+    for i, temp in enumerate(impero_c.temporaries):
+        name = "t%d" % i
+        if isinstance(temp, gem.Constant):
+            data.append(lp.TemporaryVariable(name, shape=temp.shape, dtype=temp.array.dtype, initializer=temp.array, scope=lp.temp_var_scope.GLOBAL, read_only=True))
+        else:
+            shape = tuple([i.extent for i in ctx.indices[temp]]) + temp.shape
+            data.append(lp.TemporaryVariable(name, shape=shape, dtype=numpy.float64, initializer=None, scope=lp.temp_var_scope.LOCAL, read_only=False))
+        ctx.gem_to_pymbolic[temp] = p.Variable(name)
+
+    # Create instructions
+    instructions = statement(impero_c.tree, ctx)
+
+    # Create domains
+    domains = []
+    for idx, extent in ctx.index_extent.items():
+        inames = isl.make_zero_and_vars([idx])
+        domains.append(((inames[0].le_set(inames[idx])) & (inames[idx].lt_set(inames[0] + extent))))
+
+    if not domains:
+        domains = [isl.BasicSet("[] -> {[]}")]
+
+    # Create loopy kernel
+    knl = lp.make_kernel(domains, instructions, data, name=kernel_name, target=lp.CTarget(), seq_dependencies=True, lang_version=(2018, 1))
+
+    # Prevent loopy interchange by loopy
+    knl = lp.prioritize_loops(knl, ",".join(ctx.index_extent.keys()))
+
+    # Help loopy in scheduling by assigning priority to instructions
+    insn_new = []
+    for i, insn in enumerate(knl.instructions):
+        insn_new.append(insn.copy(priority=len(knl.instructions) - i))
+    knl = knl.copy(instructions=insn_new)
+
+    return knl
+
+
+@singledispatch
+def statement(tree, ctx):
+    """Translates an Impero (sub)tree into a loopy instructions corresponding
+    to a C statement.
+
+    :arg tree: Impero (sub)tree
+    :arg ctx: miscellaneous code generation data
+    :returns: list of loopy instructions
+    """
+    raise AssertionError("cannot generate loopy from %s" % type(tree))
+
+
+@statement.register(imp.Block)
+def statement_block(tree, ctx):
+    statements = []
+    for child in tree.children:
+        statements.extend(statement(child, ctx))
+    return statements
```

Comment:
AKA `return list(statement(child, ctx) for child in tree.children)`

Reply:
Oh wait, you get a list of statements, so `list(chain(*statement(child, ctx) for child in tree.children))`

Reply:
should it be
`list(chain.from_iterable(statement(child, ctx) for child in tree.children))` ?

Reply:
Either that or `list(chain(*(statement(...) ...)))` ?

Reply:
Yep that's better.